sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -656,3 +705,301 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
+            val outerReferences = if (aggregated) {
+              collectOuterReferences(nonEquality)
+            } else {
+              AttributeSet.empty
+            }
+            val newOuterReferences = parentOuterReferences ++ outerReferences -- equivalences.keySet
+            val (newChild, joinCond, outerReferenceMap) =
+              decorrelate(child, newOuterReferences, aggregated)
+            // Add the mapping from the current node.
+            val newOuterReferenceMap = outerReferenceMap ++ equivalences
+            // Replace all outer references in non-equality filter conditions using the domain
+            // attributes produced for inner query with aggregates. This step is necessary
+            // for pushing down the non-equality filters into the domain join as join conditions.
+            val (newFilterCond, newJoinCond) = if (aggregated) {
+              val nonEqualityCond = replaceOuterReferences(nonEquality, newOuterReferenceMap)
+              (nonEqualityCond ++ uncorrelated, equality)
+            } else {
+              (uncorrelated, correlated)
+            }
+            val newFilter = newFilterCond match {
+              case Nil => newChild
+              case xs => Filter(xs.reduce(And), newChild)
+            }
+            (newFilter, joinCond ++ newJoinCond, newOuterReferenceMap)
+
+          case Project(projectList, child) =>
+            val newOuterReferences = parentOuterReferences ++ outerReferences
+            val (newChild, joinCond, outerReferenceMap) =
+              decorrelate(child, newOuterReferences, aggregated)
+            // Replace all outer references in the original project list.
+            val newProjectList = replaceOuterReferences(projectList, outerReferenceMap)
+            // Preserve required domain attributes in the join condition by adding the missing
+            // references to the new project list.
+            val referencesToAdd = missingReferences(newProjectList.map(_.toAttribute), joinCond)
+            val newProject = Project(newProjectList ++ referencesToAdd, newChild)
+            (newProject, joinCond, outerReferenceMap)
+
+          case a @ Aggregate(groupingExpressions, aggregateExpressions, child) =>
+            val newOuterReferences = parentOuterReferences ++ outerReferences
+            val (newChild, joinCond, outerReferenceMap) =
+              decorrelate(child, newOuterReferences, aggregated = true)
+            // Replace all outer references in grouping and aggregate expressions.
+            val newGroupingExpr = replaceOuterReferences(groupingExpressions, outerReferenceMap)
+            val newAggExpr = replaceOuterReferences(aggregateExpressions, outerReferenceMap)
+            // Add all required domain attributes to both grouping and aggregate expressions.
+            val groupingExprToAdd = missingReferences(newGroupingExpr, joinCond)
+            val aggExprToAdd = missingReferences(newAggExpr.map(_.toAttribute), joinCond)
```

Comment:
how do we guarantee that `aggExprToAdd` are also present in grouping expressions?

Reply:
Good catch!